processing/src/main/java/org/apache/druid/segment/join/JoinFilterAnalyzer.java

```diff
@@ -0,0 +1,743 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.join;
+
+import com.google.common.collect.ImmutableList;
+import it.unimi.dsi.fastutil.ints.IntList;
+import org.apache.druid.math.expr.Expr;
+import org.apache.druid.query.dimension.DimensionSpec;
+import org.apache.druid.query.filter.Filter;
+import org.apache.druid.query.filter.InDimFilter;
+import org.apache.druid.query.filter.ValueMatcher;
+import org.apache.druid.segment.ColumnSelectorFactory;
+import org.apache.druid.segment.ColumnValueSelector;
+import org.apache.druid.segment.DimensionSelector;
+import org.apache.druid.segment.NilColumnValueSelector;
+import org.apache.druid.segment.VirtualColumn;
+import org.apache.druid.segment.column.ColumnCapabilities;
+import org.apache.druid.segment.column.ValueType;
+import org.apache.druid.segment.filter.AndFilter;
+import org.apache.druid.segment.filter.Filters;
+import org.apache.druid.segment.filter.InFilter;
+import org.apache.druid.segment.filter.OrFilter;
+import org.apache.druid.segment.filter.SelectorFilter;
+import org.apache.druid.segment.join.lookup.LookupColumnSelectorFactory;
+import org.apache.druid.segment.join.lookup.LookupJoinable;
+import org.apache.druid.segment.join.table.IndexedTable;
+import org.apache.druid.segment.join.table.IndexedTableJoinable;
+import org.apache.druid.segment.virtual.ExpressionVirtualColumn;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+public class JoinFilterAnalyzer
+{
+  private static final String PUSH_DOWN_VIRTUAL_COLUMN_NAME_BASE = "JOIN-FILTER-PUSHDOWN-VIRTUAL-COLUMN-";
+  private static final ColumnSelectorFactory ALL_NULL_COLUMN_SELECTOR_FACTORY = new AllNullColumnSelectorFactory();
+
+  public static JoinFilterSplit splitFilter(
+      HashJoinSegmentStorageAdapter hashJoinSegmentStorageAdapter,
+      @Nullable Filter originalFilter
+  )
+  {
+    if (originalFilter == null) {
+      return new JoinFilterAnalyzer.JoinFilterSplit(
+          null,
+          null,
+          ImmutableList.of()
+      );
+    }
+
+    Filter normalizedFilter = Filters.convertToCNF(originalFilter);
+
+    // build the prefix and equicondition maps
+    Map<String, Expr> equiconditions = new HashMap<>();
+    Map<String, JoinableClause> prefixes = new HashMap<>();
+    for (JoinableClause clause : hashJoinSegmentStorageAdapter.getClauses()) {
+      prefixes.put(clause.getPrefix(), clause);
+      for (Equality equality : clause.getCondition().getEquiConditions()) {
+        equiconditions.put(clause.getPrefix() + equality.getRightColumn(), equality.getLeftExpr());
```

Comment:
There could be more than one equi-condition for the same rhs column. It looks like this code disregards that possibility.

Reply:
I'm still looking into this comment

Reply:
I tried a test where I had the following join condition:

```
"r1.regionIsoCode" == regionIsoCode && "r1.regionIsoCode" == countryIsoCode
```

In one of the Equality objects in the equiConditions, the right hand column ends up as `regionIsoCode_0` with an extra `_0` suffix, and the query fails with: 

```
"Cannot build hash-join matcher on non-key-based condition: Equality{leftExpr=countryIsoCode, rightColumn='regionIsoCode_0'}"
```
I added a new test which shows this (`JoinFilterAnalyzerTest.test_filterPushDown_factToRegionTwoColumnsToOneRHSColumnAndFilterOnRHS`)

Can we address this in a follow-on PR?

Reply:
Sure, but please raise a github issue and mention it in a comment.

Reply:
I opened https://github.com/apache/druid/issues/9327 to track the query failure issue and https://github.com/apache/druid/issues/9328 to track the adjustments needed here